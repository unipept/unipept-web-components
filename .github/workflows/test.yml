name: Test

on: [push]

jobs:
  lint:
    runs-on: ubuntu-latest
    
    steps:
    - uses: actions/checkout@v1
    - uses: actions/setup-node@v1
    - name: Install dependencies
      run: npm install
    - name: Install Jest CLI
      run: npm i -g jest-cli
    - name: Run tests
      run: jest
    
