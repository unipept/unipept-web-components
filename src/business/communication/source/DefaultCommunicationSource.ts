import CommunicationSource from "./CommunicationSource";
import EcResponseCommunicator from "@/business/communication/functional/ec/EcResponseCommunicator";
import GoResponseCommunicator from "./../functional/go/GoResponseCommunicator";
import InterproResponseCommunicator from "./../functional/interpro/InterproResponseCommunicator";
import NcbiResponseCommunicator from "./../taxonomic/ncbi/NcbiResponseCommunicator";
import Pept2DataCommunicator from "./../peptides/Pept2DataCommunicator";

export default class DefaultCommunicationSource implements CommunicationSource {
    public constructor(
        private readonly serviceUrl: string
    ) {}

    public getEcCommunicator(): EcResponseCommunicator {
        return new EcResponseCommunicator();
    }

    public getGoCommunicator(): GoResponseCommunicator {
        return new GoResponseCommunicator();
    }

    public getInterproCommunicator(): InterproResponseCommunicator {
        return new InterproResponseCommunicator();
    }

    public getNcbiCommunicator(): NcbiResponseCommunicator {
        return new NcbiResponseCommunicator();
    }

    public getPept2DataCommunicator(): Pept2DataCommunicator {
        return new Pept2DataCommunicator(this.serviceUrl);
    }
}
