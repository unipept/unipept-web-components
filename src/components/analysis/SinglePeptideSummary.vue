<template>
    <div v-if="!assay.analysisInProgress">
        <div class="display-1">
            Tryptic peptide analysis of {{ assay.peptide }}
        </div>
        <div class="subtitle-1">
            {{ assay.peptide }} was found in
            <span class="font-weight-bold">{{ proteins(assay).length }}</span>
            UniProt entries with a lowest common ancestor of
            <span class="font-weight-bold">{{ lca(assay)?.name }}</span>
        </div>
        <div class="subtitle-2">
            <span v-if="assay.equateIl">
                Equate I/L is enabled.
            </span>
            <span v-else>
                Equate I/L is disabled.
            </span>
        </div>
        <v-row>
            <v-col :cols="6">
                <div class="headline">
                    Biodiversity
                </div>
                <div v-if="lca(assay)">
                    The <span class="font-weight-bold">lowest common ancestor</span> is
                    <a
                        :href="taxonomyUrl(lca(assay))"
                        target="_blank"
                        class="link primary--text"
                    >
                        {{ lca(assay)?.name }}
                    </a> ({{ lca(assay)?.rank }}).
                </div>
                <div>
                    The <span class="font-weight-bold">common lineage</span> for all these proteins is:
                    <span v-if="commonLineage(assay).length > 0">
                        <span
                            v-for="[idx, node] of commonLineage(assay).entries()"
                            :key="idx"
                        >
                            <a
                                v-if="node"
                                :href="`https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=${node.id}`"
                                target="_blank"
                                class="link primary--text"
                            >
                                {{ node.name }}
                            </a>
                            <span v-else>
                                Unknown
                            </span>
                            <span v-if="idx + 1 < commonLineage(assay).length"> > </span>
                        </span>
                    </span>
                    <span v-else>
                        root
                    </span>
                </div>
            </v-col>
            <v-col :cols="6">
                <div class="headline">
                    Function
                </div>
                <div>
                    <span class="font-weight-bold">{{ assay.goProteinCountTableProcessor.getTrust().annotatedItems }} proteins</span>
                    ({{ toPercentage(assay.goProteinCountTableProcessor.getTrust().annotatedItems / assay.goProteinCountTableProcessor.getTrust().totalAmountOfItems) }})
                    have at least one
                    <span
                        v-if="goLink"
                        class="link primary--text"
                        @click="goLinkClicked"
                    >
                        GO term
                    </span>
                    <span
                        v-else
                        class="font-weight-bold"
                    >
                        GO term
                    </span>
                    assigned to them.
                </div>

                <div>
                    <span class="font-weight-bold">{{ assay.ecProteinCountTableProcessor.getTrust().annotatedItems }} proteins</span>
                    ({{ toPercentage(assay.ecProteinCountTableProcessor.getTrust().annotatedItems / assay.ecProteinCountTableProcessor.getTrust().totalAmountOfItems) }})
                    have at least one
                    <span
                        v-if="ecLink"
                        class="link primary--text"
                        @click="ecLinkClicked"
                    >
                        EC number
                    </span>
                    <span
                        v-else
                        class="font-weight-bold"
                    >
                        EC number
                    </span>
                    assigned to them.
                </div>

                <div>
                    <span class="font-weight-bold">
                        {{ assay.interproProteinCountTableProcessor.getTrust().annotatedItems }} proteins
                    </span>
                    ({{ toPercentage(assay.interproProteinCountTableProcessor.getTrust().annotatedItems / assay.interproProteinCountTableProcessor.getTrust().totalAmountOfItems) }})
                    have at least one
                    <span
                        v-if="interproLink"
                        class="link primary--text"
                        @click="interproLinkClicked"
                    >
                        InterPro entry
                    </span>
                    <span
                        v-else
                        class="font-weight-bold"
                    >
                        InterPro entry
                    </span>
                    assigned to them.
                </div>
            </v-col>
        </v-row>
    </div>
    <div v-else>
        <div class="display-1">
            Tryptic peptide analysis of {{ assay.peptide }}
        </div>
        <span class="subtitle-1">
            Computing summary...
        </span>
    </div>
</template>

<script setup lang="ts">
import { SinglePeptideAnalysisStatus } from '@/interface'
import { NcbiTaxon, ProteinDefinition, StringUtils } from '@/logic';

export interface Props {
    assay: SinglePeptideAnalysisStatus
    goLink: boolean
    ecLink: boolean
    interproLink: boolean
}

defineProps<Props>();

const emits = defineEmits(['goLinkClicked', 'ecLinkClicked', 'interproLinkClicked']);

const proteins = (assay: SinglePeptideAnalysisStatus): ProteinDefinition[] => {
    return assay.proteinProcessor.getProteins();
};

const lca = (assay: SinglePeptideAnalysisStatus): (NcbiTaxon | undefined) => {
    return assay.ncbiOntology.getDefinition(assay.proteinProcessor.getLca());
};

const commonLineage = (assay: SinglePeptideAnalysisStatus): (NcbiTaxon | undefined)[] => {
    return assay.proteinProcessor.getCommonLineage().map(c => assay.ncbiOntology.getDefinition(c));
};

const toPercentage = (n: number): string => {
    if (Number.isNaN(n)) {
        n = 0;
    }
    return StringUtils.numberToPercent(n);
}

const taxonomyUrl = (taxon: NcbiTaxon | undefined): string => {
    if (taxon === undefined || taxon.id === 1) {
        return 'https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi';
    }

    return `https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=${taxon.id}`;
};

const goLinkClicked = () => {
    emits('goLinkClicked');
};

const ecLinkClicked = () => {
    emits('ecLinkClicked');
};

const interproLinkClicked = () => {
    emits('interproLinkClicked');
};
</script>

<style scoped>
a {
    text-decoration: none;
}

.link {
    text-decoration: none;
    cursor: pointer;
}
</style>
