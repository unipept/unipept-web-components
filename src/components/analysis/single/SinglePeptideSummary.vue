<template>
    <div v-if="!loading">
        <div class="display-1">Tryptic peptide analysis of {{ peptide }}</div>
        <div class="subtitle-1">
            {{ peptide }} was found in
            <span class="font-weight-bold">{{ proteins.length }}</span>
            UniProt entries with a lowest common ancestor of
            <span class="font-weight-bold" v-if="lca">{{ lca.name }}</span>
        </div>
        <v-row>
            <v-col :cols="6">
                <div class="headline">Biodiversity</div>
                <div v-if="lca">
                    The <span class="font-weight-bold">lowest common ancestor</span> is
                    <a :href="`https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=${lca.id}`" target="_blank">
                        {{ lca.name }}
                    </a> ({{ lca.rank }}).
                </div>
                <div>
                    The <span class="font-weight-bold">common lineage</span> for all these proteins is:
                    <span v-if="commonLineage.length > 0">
                        <span v-for="[idx, node] of commonLineage.entries()" :key="idx">
                            <a v-if="node" :href="`https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=${node.id}`" target="_blank">
                                {{ node.name }}
                            </a>
                            <span v-else>
                                Unknown
                            </span>
                            <span v-if="idx + 1 < commonLineage.length"> > </span>
                        </span>
                    </span>
                    <span v-else>
                        root
                    </span>
                </div>
            </v-col>
            <v-col :cols="6">
                <div class="headline">Function</div>
                <div v-if="goTrust">
                    <span class="font-weight-bold">{{ goTrust.annotatedItems }} proteins</span>
                    ({{ toPercentage(goTrust.annotatedItems / goTrust.totalAmountOfItems) }})
                    have at least one <span class="font-weight-bold">GO term</span> assigned to them.
                </div>
                <div v-if="ecTrust">
                    <span class="font-weight-bold">{{ ecTrust.annotatedItems }} proteins</span>
                    ({{ toPercentage(ecTrust.annotatedItems / ecTrust.totalAmountOfItems) }})
                    have at least one <span class="font-weight-bold">EC number</span> assigned to them.
                </div>
                <div v-if="interproTrust">
                    <span class="font-weight-bold">{{ interproTrust.annotatedItems }} proteins</span>
                    ({{ toPercentage(interproTrust.annotatedItems / interproTrust.totalAmountOfItems) }})
                    have at least one <span class="font-weight-bold">InterPro entry</span> assigned to them.
                </div>
            </v-col>
        </v-row>
    </div>
    <div v-else>
        <div class="display-1">Tryptic peptide analysis of {{ peptide }}</div>
        <span class="subtitle-1">Computing summary...</span>
    </div>
</template>

<script lang="ts">
import Vue from "vue";
import Component from "vue-class-component";
import { Peptide } from "./../../../business/ontology/raw/Peptide";
import { Prop, Watch } from "vue-property-decorator";
import ProteinDefinition from "./../../../business/ontology/protein/ProteinDefinition";
import FunctionalTrust from "./../../../business/processors/functional/FunctionalTrust";
import StringUtils from "./../../../business/misc/StringUtils";
import NcbiTaxon from "./../../../business/ontology/taxonomic/ncbi/NcbiTaxon";
import ProteinProcessor from "./../../../business/processors/protein/ProteinProcessor";
import GoProteinCountTableProcessor from "./../../../business/processors/functional/go/GoProteinCountTableProcessor";
import EcProteinCountTableProcessor from "./../../../business/processors/functional/ec/EcProteinCountTableProcessor";
import InterproProteinCountTableProcessor
    from "./../../../business/processors/functional/interpro/InterproProteinCountTableProcessor";
import NcbiOntologyProcessor from "./../../../business/ontology/taxonomic/ncbi/NcbiOntologyProcessor";
import CommunicationSource from "./../../../business/communication/source/CommunicationSource";

@Component
export default class SinglePeptideSummary extends Vue {
    @Prop({ required: true })
    private peptide: Peptide;
    @Prop({ required: true })
    private equateIl: boolean;
    @Prop({ required: true })
    private communicationSource: CommunicationSource;

    private proteins: ProteinDefinition[] = [];
    private lca: NcbiTaxon = null;
    private commonLineage: NcbiTaxon[] = [];

    private goTrust: FunctionalTrust = null;
    private ecTrust: FunctionalTrust = null;
    private interproTrust: FunctionalTrust = null;

    private loading: boolean = false;

    private mounted() {
        this.onInputsChanged();
    }

    @Watch("peptide")
    @Watch("equateIl")
    private async onInputsChanged() {
        if (this.peptide) {
            this.loading = true;
            const proteinProcessor = new ProteinProcessor();
            this.proteins = await proteinProcessor.getProteinsByPeptide(this.peptide, this.equateIl);

            const goProcessor = new GoProteinCountTableProcessor(this.peptide, this.equateIl);
            this.goTrust = await goProcessor.getTrust();

            const ecProcessor = new EcProteinCountTableProcessor(this.peptide, this.equateIl);
            this.ecTrust = await ecProcessor.getTrust();

            const interproProcessor = new InterproProteinCountTableProcessor(this.peptide, this.equateIl);
            this.interproTrust = await interproProcessor.getTrust();

            const lca = await proteinProcessor.getLcaByPeptide(this.peptide, this.equateIl);
            const commonLineage = await proteinProcessor.getCommonLineageByPeptide(this.peptide, this.equateIl);

            const ncbiOntologyProcessor = new NcbiOntologyProcessor(this.communicationSource);
            const ontology = await ncbiOntologyProcessor.getOntologyByIds([lca, ...commonLineage]);

            this.lca = ontology.getDefinition(lca);

            this.commonLineage.length = 0;
            this.commonLineage.push(...commonLineage.map(c => ontology.getDefinition(c)));

            this.loading = false;
        }
    }

    private toPercentage(n: number): string {
        if (Number.isNaN(n)) {
            n = 0;
        }
        return StringUtils.numberToPercent(n);
    }
}
</script>

<style scoped>

</style>
