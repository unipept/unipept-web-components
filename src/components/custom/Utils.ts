export default class Utils {
    /**
     * This method should be used when a specific URL should be opened in a new browser window. The method automatically
     * decides whether Electron or a default redirection should take place.
     * 
     * @param url The full url to which navigation should take place.
     */
    static openInBrowser(url: string): void {
        if (Utils.isElectron()) {
            const shell = require("electron").shell;
            shell.openExternal(url);
        } else {
            window.open(url);
        }
    }
    
    /**
     * Method that can be used to check if the app is being executed by Electron or something else.
     * 
     * @return True if the app is being executed in an Electron-powered environment.
     */
    static isElectron(): boolean {
        const userAgent = navigator.userAgent.toLowerCase();
        return userAgent.indexOf(" electron/") > -1;
    }
}
