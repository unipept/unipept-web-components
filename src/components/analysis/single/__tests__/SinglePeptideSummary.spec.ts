import Vue from "vue"
import Vuetify from "vuetify"
import { createLocalVue, mount } from "@vue/test-utils";
import Setup from "@/test/Setup";
import SinglePeptideSummary from "@/components/analysis/single/SinglePeptideSummary.vue";
import { waitForElement } from "@/test/Utils";
import DefaultCommunicationSource from "@/business/communication/source/DefaultCommunicationSource";

Vue.use(Vuetify);

const localVue = createLocalVue();

describe("SinglePeptideSummary", () => {
    let vuetify;

    beforeEach(() => {
        vuetify = new Vuetify();
    });

    beforeAll(() => {
        const setup = new Setup();
        setup.setupAll();
    });

    it("correctly shows a complete summary for one peptide", async() => {
        const wrapper = mount(SinglePeptideSummary, {
            localVue,
            vuetify,
            propsData: {
                peptide: "AALTER",
                equateIl: true,
                communicationSource: new DefaultCommunicationSource("http://unipept.ugent.be")
            }
        });

        await wrapper.vm.$nextTick();

        await waitForElement(wrapper, ".headline");

        expect(wrapper.html()).toMatchSnapshot();
    });
});
