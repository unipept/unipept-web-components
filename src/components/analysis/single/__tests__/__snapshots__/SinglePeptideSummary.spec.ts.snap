// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SinglePeptideSummary correctly shows a complete summary for one peptide 1`] = `
"<div><div class=\\"display-1\\">Tryptic peptide analysis of AALTER</div> <div class=\\"subtitle-1\\">
        AALTER was found in
        <span class=\\"font-weight-bold\\">8</span>
        UniProt entries with a lowest common ancestor of
        <span class=\\"font-weight-bold\\">root</span></div> <div class=\\"row\\"><div class=\\"col col-6\\"><div class=\\"headline\\">Biodiversity</div> <div>
                The <span class=\\"font-weight-bold\\">lowest common ancestor</span> is
                <a href=\\"https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&amp;id=1\\">
                    root
                </a> (no rank).
            </div> <div>
                The <span class=\\"font-weight-bold\\">common lineage</span> for all these proteins is:
                <span>
                    root
                </span></div></div> <div class=\\"col col-6\\"><div class=\\"headline\\">Function</div> <div><span class=\\"font-weight-bold\\">8 proteins</span>
                (100%)
                have at least one <span class=\\"font-weight-bold\\">GO term</span> assigned to them.
            </div> <div><span class=\\"font-weight-bold\\">7 proteins</span>
                (88%)
                have at least one <span class=\\"font-weight-bold\\">EC number</span> assigned to them.
            </div> <div><span class=\\"font-weight-bold\\">8 proteins</span>
                (100%)
                have at least one <span class=\\"font-weight-bold\\">InterPro entry</span> assigned to them.
            </div></div></div></div>"
`;
